package/src/error.rs

```diff
@@ -0,0 +1,188 @@
+        /// The git id of the bad package.
+        package_commit: ObjectId,
+        /// The relative path of the bad package within its git repo.
+        package_path: PathBuf,
+        attempted: PathBuf,
+        restriction: PathBuf,
+    },
+    /// There was some error interacting with a git repository.
+    Git(nickel_lang_git::Error),
+    InvalidUrl {
+        url: String,
+        msg: String,
+    },
+    Resolution {
+        msg: String,
+    },
+    InternalManifestError {
+        path: PathBuf,
+        msg: String,
+    },
+    /// There was an error persisting a temporary file.
+    TempFilePersist {
+        error: tempfile::PersistError,
+    },
+    NoManifestParent,
+    /// Index dependencies aren't implemented yet, so we emit
+    /// this if we encounter one.
+    IndexDep,
+}
+
+impl std::error::Error for Error {}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::Io { error, path } => {
+                if let Some(path) = path {
+                    write!(f, "{}: {error}", path.display())
+                } else {
+                    error.fmt(f)
+                }
+            }
+            // Just a short and not-very-informative error. To write a better error message to
+            // the terminal, use `program.report` like the cli does.
+            Error::ManifestEval { package, .. } => {
+                if let Some(package) = package {
+                    write!(f, "error evaluating manifest for package {package}")
+                } else {
+                    write!(f, "error evaluating package manifest")
+                }
+            }
+            Error::RestrictedPath {
+                attempted,
+                restriction,
+                package_url,
+                package_commit,
+                package_path,
+            } => {
+                write!(
+                    f,
+                    "git package {package_url}@{package_commit}/{} tried to import path {}, but can only import from {}",
+                    package_path.display(),
+                    attempted.display(),
+                    restriction.display()
+                )
+            }
+            Error::Git(e) => e.fmt(f),
+            Error::InvalidUrl { msg, .. } => {
+                write!(f, "{}", msg)
+            }
+            Error::InternalManifestError { path, msg } => {
+                write!(
+                    f,
+                    "internal error reading the manifest at {}; this is a bug in nickel: {msg}",
+                    path.display()
+                )
+            }
+            Error::NoPackageRoot { path } => write!(
+                f,
+                "tried to import a relative path ({}), but we have no reference",
```

Comment:
I'm not sure I would understand this error message without context.

Reply:
Addendum: I think "reference" wasn't clear initially. I find "root" better, or "base" maybe, for what it's worth

Reply:
I've gone with root for now. I have a TODO to enrich these sorts of errors with dependency chains, which I think will make them easier to understand.